---
title: "Decoil reconstruction results"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    keep_md: TRUE
knitr:
  opts_chunk:
    echo: true
    message: true
    warning: true
params:
  help: false
  maxcov: 600
  full: TRUE
  refgenome: "/anno/reference.fa"
  gtffile: "/anno/anno.gtf"
  genesfile: "/code/anno/genes.txt"
  root: "/example/"
  fname: "reconstruct.test"
  sample: "test"
  bw: "/example/coverage.bw"
  nodesfile: "/example/reconstruct.ecDNA.filtered.bed"
  edgesfile: "/example/reconstruct.links.ecDNA.filtered.txt"
  summaryfile: "/example/summary.txt"
  window: ""
  decoilvizversion: "1.0.4"
  filtertop: ""
  filterscore: ""
  extendallowedchr: "chr1,chr2"
fontsize: 9pt
graphics: yes

---

```{css toc-content, echo = FALSE}
#TOC {
  background-color: #165d92;
  margin-top: 200px;
}

.level3 {
  margin: 1%;
  padding-top: 20px;
  padding-left: 300px;
}

.toc-content {
    width: 100%;
    margin: 0px;
    padding: 0px;
}

h1, h2, h3, h4, h5, h6 {
    margin-top: 20px;
    margin-bottom: 30px;
    font-weight: 400;
    color: #165d92;
}

.title {
    display: inline-block;
    width: 110%;
    color: #fff;
    margin:0;
    text-align: center;
    background-color: #159957;
    background-image: linear-gradient(120deg,#155799,#159957);
    padding: 50px;
}

.main-container {

    padding: 0px;
    width: 90%;
    margin: auto;
    max-width: 2000px;
}

```

```{r setup, echo=FALSE, include=FALSE, message = FALSE, warning=FALSE}
knitr::opts_chunk$set(ev = "png",
                      dpi = 600,
                      echo = FALSE,
                      cache = FALSE,
                      fig.width=7,
                      out.width="100%",
                      root.dir = params$root)
library(rtracklayer)
library(pafr)
library(gTrack)
library(gGnome)
library(GenomicRanges)
library(dplyr)
library(rmarkdown)
library(knitr)
library(kableExtra)
library(Gviz)
library(plyranges)
library(assertthat)
library(DT)

root <- params$root
setwd(root)

nodesfile <- params$nodesfile
edgesfile <- params$edgesfile
summfile <- params$summaryfile
bwfile <- params$bw
sample <- params$sample
maxcov <- params$maxcov
fname <- params$fname
htmlfile <- paste0(root, fname, ".html")

decoilvizversion <- params$decoilvizversion
refgenome <- params$refgenome
gtffile <- params$gtffile
genes <- params$genesfile
windowfile <- params$window

filtertop <- params$filtertop
filterscore <- params$filterscore

allowedchr <- unlist(strsplit(params$extendallowedchr, split=","))

sc = "#171717"
sv = "#226258"
msv = "#3D2C2E"
mic = "#F4D35E"
mrinter = "#E4644E"
sdup = "#13908E"
fb = "#613C68"

colors <- c(sc, sv, msv, mic, mrinter, sdup, fb)
corder = c("Simple circle", "Simple SV's", "Mixed SV's", "Multi-region intra-chromosomal", 
          "Multi-region inter-chromosomal","Simple duplications","Foldbacks")

```

### Configuration

>**Date**: `r Sys.Date()` <br/>
>**Decoil-viz version**: `r decoilvizversion`

>**Reference genome**: `r refgenome` <br/>
>**GTF annotation file**: `r gtffile` <br/>
>**Extend allowed chromosome annotation**: `r allowedchr` <br/>
>**Gene list**: `r genes` 


>**Sample name**: `r sample` <br/>
>**Genomic regions**: `r nodesfile` <br/>
>**Breakpoints information**: `r edgesfile` <br/>
>**Coverage track**: `r bwfile` <br/>
>**ROI (region of interest)**: `r windowfile` <br/>
>**Output**: `r htmlfile`
<br/>

```{r assert_paths, echo=FALSE, message = FALSE, warning=FALSE, results='asis', cache = FALSE}

# check paths
library(assertthat)
assert_that(file.exists(edgesfile), msg = paste0("edges file exists ", edgesfile))
assert_that(file.exists(nodesfile), msg = paste0("nodes file exists ", nodesfile))
assert_that(file.exists(summfile), msg = paste0("summary file exists ", summfile))
assert_that(file.exists(bwfile), msg = paste0("coverage file exists ", bwfile))
assert_that(file.exists(refgenome), msg = paste0("refgenome file exists ", refgenome))
assert_that(file.exists(gtffile), msg = paste0("gtffile file exists ", gtffile))

# if window file specified please check length
if (length(windowfile) > 1){
  assert_that(file.exists(windowfile), msg = paste0("window file exists", windowfile))
}

# check if data available
ed1<-readLines(edgesfile)
if (length(ed1) < 2){
  rtext <- kableExtra::text_spec("No ecDNA elements / cycles found", "html", bold = TRUE, color = "red", font_size = 20)
  knitr::knit_exit()
}else{
  df_edges <- read.csv2(edgesfile,header=T,sep="\t")
  if (dim(df_edges)[1] == 0){
    rtext <- kableExtra::text_spec("No ecDNA elements / cycles found", "html", bold = TRUE, color = "red", font_size = 20)
    cat(rtext)
    knitr::knit_exit()
  }
}

```

```{r load_annotation, echo=FALSE, message=FALSE, warning=FALSE, results='asis', cache=FALSE}
#### 1. Load genome annotation 

# create the track.gencode plot

# allowed chromosome annotation
primary_chromosomes <-
  c(
    "chr1",
    "chr2",
    "chr3",
    "chr4",
    "chr5",
    "chr6",
    "chr7",
    "chr8",
    "chr9",
    "chr10",
    "chr11",
    "chr12",
    "chr13",
    "chr14",
    "chr15",
    "chr16",
    "chr17",
    "chr18",
    "chr19",
    "chr20",
    "chr21",
    "chr22",
    "chrX",
    "chrY",
    "chrM",
    "MT",
    "M",
    "X",
    "Y",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "9",
    "10",
    "11",
    "12",
    "13",
    "14",
    "15",
    "16",
    "17",
    "18",
    "19",
    "20",
    "21",
    "22",
    "CP068255.2",
    "CP068256.2",
    "CP068257.2",
    "CP068258.2",
    "CP068259.2",
    "CP068260.2",
    "CP068261.2",
    "CP068262.2",
    "CP068263.2",
    "CP068264.2",
    "CP068265.2",
    "CP068266.2",
    "CP068267.2",
    "CP068268.2",
    "CP068269.2",
    "CP068270.2",
    "CP068271.2",
    "CP068272.2",
    "CP068273.2",
    "CP068274.2",
    "CP068275.2",
    "CP068276.2",
    "CP068277.2",
    "CP086569.2",
    "CP068254.1",
    "NC_060925.1",
    "NC_060926.1",
    "NC_060927.1",
    "NC_060928.1",
    "NC_060929.1",
    "NC_060930.1",
    "NC_060931.1",
    "NC_060932.1",
    "NC_060933.1",
    "NC_060934.1",
    "NC_060935.1",
    "NC_060936.1",
    "NC_060937.1",
    "NC_060938.1",
    "NC_060939.1",
    "NC_060940.1",
    "NC_060941.1",
    "NC_060942.1",
    "NC_060943.1",
    "NC_060944.1",
    "NC_060945.1",
    "NC_060946.1",
    "NC_060947.1",
    "NC_060948.1",
    "NC_000001.11",
    "NC_000002.12",
    "NC_000003.12",
    "NC_000004.12",
    "NC_000005.10",
    "NC_000006.12",
    "NC_000007.14",
    "NC_000008.11",
    "NC_000009.12",
    "NC_000010.11",
    "NC_000011.10",
    "NC_000012.12",
    "NC_000013.11",
    "NC_000014.9",
    "NC_000015.10",
    "NC_000016.10",
    "NC_000017.11",
    "NC_000018.10",
    "NC_000019.10",
    "NC_000020.11",
    "NC_000021.9",
    "NC_000022.11",
    "NC_000023.11",
    "NC_000024.10",
    "NC_012920.1",
    "NC_000067.7",
    "NC_000068.8",
    "NC_000069.7",
    "NC_000070.7",
    "NC_000071.7",
    "NC_000072.7",
    "NC_000073.7",
    "NC_000074.7",
    "NC_000075.7",
    "NC_000076.7",
    "NC_000077.7",
    "NC_000078.7",
    "NC_000079.7",
    "NC_000080.7",
    "NC_000081.7",
    "NC_000082.7",
    "NC_000083.7",
    "NC_000084.7",
    "NC_000085.7",
    "NC_000086.8",
    "NC_000087.8",
    "NC_005089.1"
  )

# add additional chromosome annotation
primary_chromosomes <- union(primary_chromosomes,allowedchr)

# import gtf
gr <- rtracklayer::import(gtffile)

filtered_gr <- plyranges::filter(gr, seqnames %in% primary_chromosomes &
                        gene_name %in% read.csv2(genes,header=F,"\t")$V1 &
                        type %in% c('gene'))

keep <- !is.na(mcols(filtered_gr)$gene_id)
filtered_gr <- filtered_gr[keep,]

alllevels <- seqlevels(filtered_gr)
filteredlevels <- alllevels[alllevels %in% primary_chromosomes]
seqlevels(filtered_gr) <- filteredlevels

# generate gencode track and make sure it has the correct attributes/labelfields (gene, gene_name)
namecols <- colnames(mcols(filtered_gr))
if ('gene_name' %in% namecols){
  gencode = gTrack(filtered_gr, name = 'Annotation', gr.labelfield = 'gene_name', grl.labelfield = 'gene_name', height = 90, stack.gap = 1e+06, col=alpha('#159957', .5), xaxis.round = NA, circles = FALSE)
}else if ('gene' %in% namecols){
  gencode = gTrack(filtered_gr, name = 'Annotation', gr.labelfield = 'gene', grl.labelfield = 'gene', height = 90, stack.gap = 1e+06, col=alpha('#159957', .5), xaxis.round = NA, circles = FALSE)
}else{
  print("GTF file is not correct. Make sure you have gene_name or gene as attribute")
  rtext <- kableExtra::text_spec("GTF file is not correct. Make sure you have gene_name or gene as attribute", "html", bold = TRUE, color = "red", font_size = 20)
  knitr::knit_exit()
}

```

> **Available chromosome annotation:** `r primary_chromosomes`

```{r read_nodes_edges, echo=TRUE, include=FALSE, message = FALSE, warning=FALSE}
#### 2. Read nodes

df_nodes <- read.csv2(nodesfile,header=T,sep="\t")
gr_nodes <- makeGRangesFromDataFrame(df_nodes,
                                     keep.extra.columns=TRUE,
                                     ignore.strand=FALSE,
                                     seqinfo=NULL,
                                     seqnames.field=c("X.chr","seqnames", "seqname",
                                                      "chromosome", "chrom",
                                                      "chr", "chromosome_name",
                                                      "seqid"),
                                     start.field="start",
                                     end.field=c("end", "stop"),
                                     strand.field="strand",
                                     starts.in.df.are.0based=FALSE)

#### 2.1 Create gencode gTrack
# inter_gr <- GenomicRanges::intersect(filtered_gr, gr_nodes)
# gencode = gTrack(filtered_gr, name = 'Gencode', gr.labelfield = 'gene_name', gr.cex.label = 'gene_name')

#### 3. Read edges
df_edges <- read.csv2(edgesfile,header=T,sep="\t")

#### 3.1 Max coverage
if (!is.na(maxcov) && maxcov != "" && (maxcov)>0){
  maxcov <- max(maxcov, 2*max(df_nodes$coverage))
}else{
  maxcov <- 2*max(df_nodes$coverage)
}

### 3.2 Read summary
df_summary <- read.csv2(summfile,header=T,sep="\t")

### 3.3 Filtered data

# # filter nodes top X set
# if (filtertop != ''){
#   circ_list <- unique(df_nodes$circ_id)
#   circ_short_list <- circ_list[1:as.numeric(filtertop)]
#   print(circ_short_list)
#   df_nodes <- df_nodes[df_nodes$circ_id %in% circ_short_list,]
#   # df_summary <- df_summary[df_summary$circ_id %in% circ_short_list,]
#   df_edges <- df_edges[df_edges$circ_id %in% circ_short_list,]
# }
# 
# # filter nodes with a score
# if (filterscore != ''){
#   df_nodes <- df_nodes[df_nodes$estimated_proportions >= as.numeric(filterscore),]
#   # df_summary <- df_summary[df_summary$estimated_proportions >= filterscore,]
#   circ_list <- unique(df_nodes$circ_id)
#   df_edges <- df_edges[df_edges$circ_id %in% circ_list,]
# }
# 
# print(df_nodes)
```

### Reconstruction of the invidiual ecDNA elements
<br/>

We display here the individual ecDNA elements.<br/>
The visualization contains from top to bottom the following tracks:<br/>

- **Circ_id**: 
  - thread reconstruction, where the fragments are marked black and the connecting breakpoints red
  - track name is the id of the ecDNA element (found in the table as the circ_id)
- **Coverage**: coverage profile of all the reads aligning to the reference genome
- **Gencode**: gene annotation (includes only protein-coding genes)

```{r draw_amplicons_function, echo=FALSE, warning=FALSE,  fig.height=8, results='asis', out.width="100%"}

get_gradient <- function(start,end){
  colfunc <- colorRampPalette(c("gray", "black"))
  return(colfunc(end-start))
}

draw_legend_topology <- function(){
  # plot.new()
  legend("topright", legend = corder, col = colors, lty = 1, lwd = 5)
}

#### 4. Draw circles
draw_amplicon <- function(s, bwpath, maxcov, df_edges, gr_nodes, df_summary, windowfile, root){

  gg_all <- c()
  plots_all <- c()
  gRangesCov <- import(bwpath)
  gTrackCov <- gTrack(gRangesCov,
                      y.field="score",
                      line=TRUE,
                      name = "Coverage",
                      y0=0,
                      y1=maxcov,
                      height=60)

  individual_circles <- unique(df_edges$circ_id)
  max_proportion <- max(df_summary[df_summary$label == "ecDNA",]$estimated_proportions)
  min_proportion <- min(df_summary[df_summary$label == "ecDNA",]$estimated_proportions)

  current_nrnodes <- 0
  for( circ in individual_circles){

    nodes1 <- gr_nodes[(elementMetadata(gr_nodes)[, "circ_id"] == circ)]
    circ_ep <- df_summary[df_summary$circ_id==circ,]$estimated_proportions
    
    gradient_palette <- get_gradient(min_proportion, max_proportion)
    nodes1$col <- alpha(gradient_palette[circ_ep-min_proportion+1], .7)
    
    edges1 <- df_edges[df_edges$circ_id==circ,]
    edge_color <- df_summary[df_summary$circ_id==circ,]$topology_idx
    edges1$col <- alpha(colors[edge_color+1], .8)
    
    fcirc <- paste0(fname, "/", circ, ".pdf")

    cat("Reconstruction for sample:",s,"<br/>")
    cat("EcDNA element: circ_id =", circ, "<br/>")
    # cat("<a href=\"", fcirc,"\">Download circ_id =",circ,"</a></br/>")
    # cat("Estimated proportions: ", nodes1$estimated_proportions[1,], "<br/>")

    # reset nodes numbering
    rownames(edges1) <- NULL
    edges1$n1 <- edges1$n1 - current_nrnodes
    edges1$n2 <- edges1$n2 - current_nrnodes

    gg1 = gG(nodes = nodes1,
             edges = edges1,
             meta = list(sep.bg.col=alpha("white",0)))

    # adjust the hight of the plot depending of the ecDNA complexity
    if (dim(edges1)[1]<=2){
      gg1$set(name = circ, height = 20)
    }else{
      gg1$set(name = circ, height = 50)
    }

    gt_temp <- gg1$gt
    gt_temp$sep.bg.col <- NA
    
    # window interval specified
    if (length(windowfile)>1){
      windows <- rtracklayer::import(windowfile)
      windows <- copy(windows) + 200000
      windows <- reduce(windows)
    }else{
      windows <- copy(nodes1) + 200000
      windows <- reduce(windows)
    }
    

    if (length(gg_all) > 0){
      gg_all <- c(gg_all, gt_temp)
      plots_all <- c(plots_all, fcirc)
    }else{
      gg_all <- c(gt_temp)
      plots_all <- c(fcirc)
    }
    current_nrnodes <- current_nrnodes + length(nodes1)

    # plot as png
    gTrack::plot.gTrack(c(gencode, gTrackCov, gt_temp),
                        windows, sep.draw = FALSE)
    
    # plot as pdf
    pdf(paste0(root, fcirc))
    gTrack::plot.gTrack(c(gencode, gTrackCov, gt_temp),
                        windows,sep.draw = FALSE)
    dev.off()
    
    cat("<br/>")
    
  }

  return(list("gplots"=gg_all, "plotsall"=plots_all))

}

```

```{r draw_amplicons, echo=FALSE, warning=FALSE,  fig.height=8, results='asis', out.width="100%"}
#### 5. Draw individuals 
out <- draw_amplicon(sample, bwfile, maxcov, df_edges, gr_nodes, df_summary, windowfile, root)
p1 <- out$gplots
paths_plots <- out$plotsall
```
### Summary of all found ecDNA elements
<br/>

```{r draw_amplicons_summary, echo=FALSE, warning=FALSE,  fig.height=9, results='asis', out.width="100%"}
if (length(windowfile)>1){
  windows <- rtracklayer::import(windowfile)
  windows <- copy(windows) + 200000
  windows <- reduce(windows)
}else{
  windows <- copy(gr_nodes) + 200000
  windows <- reduce(windows)
}

gRangesCov <- import(bwfile)
gTrackCov <- gTrack(gRangesCov,
                    y.field="score",
                    line=TRUE,
                    name = "Coverage",
                    y0=0, y1=maxcov, height=70)

# as pdf
fcirc <- paste0(fname, "/all_ecdna_elements.pdf")
paths_plots <- c(paths_plots, fcirc)

pdf(paste0(root,fcirc))
gTrack::plot.gTrack(c(gencode, gTrackCov,p1),
                    windows,
                    sep.draw = FALSE)
dev.off()

# as png
gTrack::plot.gTrack(c(gencode, gTrackCov,p1),
                    windows,
                    sep.draw = FALSE)
draw_legend_topology()
```

#### Download as pdf:

```{r show_links,  echo=FALSE, warning=FALSE,  fig.height=6, results='asis', out.width="100%"}
for (pi in paths_plots){
  cat("<a href=\"", pi,"\" target=\"_blank\">",pi,"</a></br/>")
}
```


### Summary of found ecDNA elements / cycles
<br/>

We summarize below all the found ecDNA elements / cycles.<br/>
Header description:<br/>

- **circ_id** - id of the ecDNA element / cycle
- **chr_origin** - chromosomes 
- **size.MB** - sequence length of the ecDNA element / cycle
- **label** - annotation; is the cycle an ecDNA element
- **topology_idx** - index of the individual topology
- **topology_name** - topology name 
- **estimated_proportions** - copy-number estimates of the ecDNA element; the higher the more confident is the reconstruction

```{r show_summary, echo=FALSE, out.width="100%", results='asis'}
df_summary <- read.csv2(summfile,header=T,sep="\t") 
df_intersect <- filter(df_summary, circ_id %in% df_nodes$circ_id)

library(DT)
datatable(df_intersect)
```


We summarize below all the fragments composing all the ecDNA elements / cycles.<br/>
Header description:<br/>

- **X.chr** - chromosome
- **start** - start of contiguous genomic region composing the ecDNA element / cycle
- **end** - end of contiguous genomic region composing the ecDNA element / cycle
- **circ_id** - id of the ecDNA element / cycle
- **fragment_id** - unique identifier concatenation of the genomic non-overlapping fragment(s) which describe a contiguous genomic region
- **strand** - orientation of the contiguous genomic region
- **coverage** - mean coverage over the contiguous genomic region
- **estimated_proportions** - copy-number estimates of the ecDNA element / cycle; the higher the more confident is the reconstruction


```{r show_summary_fragments, echo=FALSE, out.width="100%", results='asis'}
# library(rmarkdown)
# paged_table(df_nodes)
library(DT)
datatable(df_nodes)
```

###  ROI 
Region of interest to display the reconstructions (showed only if parameter set)
```{r show_target, echo=FALSE, out.width="100%", results='asis'}
library(DT)
if (length(windowfile)>1){
  window <- read.csv(windowfile,sep="\t",header=T)
  datatable(window)
}
```

### R session information

```{r}
sessionInfo()
```